File: app_api/models/subscriptions.py
from datetime import datetime

from pydantic import BaseModel, Field


class SubscriptionCreate(BaseModel):
    """Request body for subscribing to a playlist."""
    playlist_uuid: str = Field(..., description="UUID of the playlist to subscribe to")


class SubscriptionUpdate(BaseModel):
    """Request body for updating a subscription."""
    notifications_enabled: bool | None = None


class Subscription(BaseModel):
    """Full subscription model returned from database."""
    id: str
    user_id: str
    playlist_uuid: str
    notifications_enabled: bool
    created_at: datetime
    updated_at: datetime

    # Joined data from playlists table
    playlist_name: str | None = None
    playlist_descr: str | None = None

    # Computed counts
    keyword_group_count: int = 0

    # Transformed fields for frontend (camelCase)
    userId: str | None = None
    playlistUuid: str | None = None
    notificationsEnabled: bool | None = None
    createdAt: str | None = None
    updatedAt: str | None = None
    playlistName: str | None = None
    playlistDescr: str | None = None
    keywordGroupCount: int | None = None

    class Config:
        from_attributes = True


class SubscriptionSummary(BaseModel):
    """Minimal subscription info for lists."""
    id: str
    playlist_uuid: str
    playlist_name: str | None = None
    notifications_enabled: bool
    keyword_group_count: int = 0

    # Transformed fields for frontend
    playlistUuid: str | None = None
    playlistName: str | None = None
    notificationsEnabled: bool | None = None
    keywordGroupCount: int | None = None

    class Config:
        from_attributes = True


class SubscriptionListResponse(BaseModel):
    """Response for listing user's subscriptions."""
    subscriptions: list[SubscriptionSummary]
    total: int


class SubscriptionStatus(BaseModel):
    """Quick status check for a playlist (is user subscribed?)."""
    playlist_uuid: str
    is_subscribed: bool
    subscription_id: str | None = None

    # Transformed fields for frontend
    playlistUuid: str | None = None
    isSubscribed: bool | None = None
    subscriptionId: str | None = None

    class Config:
        from_attributes = True


class LinkKeywordGroupRequest(BaseModel):
    """Request body for linking a keyword group to a subscription."""
    keyword_group_id: str = Field(..., description="UUID of the keyword group to link")


class LinkedKeywordGroup(BaseModel):
    """Keyword group linked to a subscription."""
    id: str
    keyword_group_id: str
    keyword_group_name: str
    keyword_count: int = 0
    created_at: datetime

    # Transformed fields for frontend
    keywordGroupId: str | None = None
    keywordGroupName: str | None = None
    keywordCount: int | None = None
    createdAt: str | None = None

    class Config:
        from_attributes = True

File: frontend/src/api/subscriptions.ts
/**
 * Subscriptions API module
 */

import { api, isMock } from '@/lib/api';
import type {
  Subscription,
  SubscriptionSummary,
  SubscriptionListResponse,
  SubscriptionStatus,
  SubscriptionCreate,
  SubscriptionUpdate,
  LinkedKeywordGroup,
  LinkKeywordGroupRequest,
} from '@/types/subscription';

// Mock data for development
const mockSubscriptions: SubscriptionSummary[] = [
  {
    id: 'sub-1',
    playlistUuid: 'playlist-1',
    playlistName: 'Prosper PD',
    notificationsEnabled: true,
    keywordGroupCount: 2,
  },
  {
    id: 'sub-2',
    playlistUuid: 'playlist-2',
    playlistName: 'Frisco PD',
    notificationsEnabled: true,
    keywordGroupCount: 1,
  },
];

/**
 * List all subscriptions for the current user
 */
export async function listSubscriptions(): Promise<SubscriptionListResponse> {
  if (isMock()) {
    return { subscriptions: mockSubscriptions, total: mockSubscriptions.length };
  }

  try {
    const response = await api.get<SubscriptionListResponse>('/subscriptions');
    return response.data;
  } catch (error) {
    console.warn('Using mock subscriptions due to API error', error);
    return { subscriptions: mockSubscriptions, total: mockSubscriptions.length };
  }
}

/**
 * Get subscription status for a playlist
 */
export async function getSubscriptionStatus(playlistUuid: string): Promise<SubscriptionStatus> {
  if (isMock()) {
    const sub = mockSubscriptions.find(s => s.playlistUuid === playlistUuid);
    return {
      playlistUuid,
      isSubscribed: !!sub,
      subscriptionId: sub?.id,
    };
  }

  try {
    const response = await api.get<SubscriptionStatus>(`/subscriptions/status/${playlistUuid}`);
    return response.data;
  } catch (error) {
    console.warn('Using mock subscription status due to API error', error);
    return { playlistUuid, isSubscribed: false };
  }
}

/**
 * Get a single subscription with details
 */
export async function getSubscription(subscriptionId: string): Promise<Subscription> {
  if (isMock()) {
    const sub = mockSubscriptions.find(s => s.id === subscriptionId);
    if (!sub) throw new Error('Subscription not found');
    return {
      ...sub,
      userId: 'user-1',
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };
  }

  const response = await api.get<Subscription>(`/subscriptions/${subscriptionId}`);
  return response.data;
}

/**
 * Subscribe to a playlist
 */
export async function createSubscription(data: SubscriptionCreate): Promise<Subscription> {
  if (isMock()) {
    return {
      id: `sub-${Date.now()}`,
      userId: 'user-1',
      playlistUuid: data.playlist_uuid,
      notificationsEnabled: true,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
      keywordGroupCount: 0,
    };
  }

  const response = await api.post<Subscription>('/subscriptions', data);
  return response.data;
}

/**
 * Update subscription settings
 */
export async function updateSubscription(
  subscriptionId: string,
  data: SubscriptionUpdate
): Promise<Subscription> {
  if (isMock()) {
    const sub = mockSubscriptions.find(s => s.id === subscriptionId);
    if (!sub) throw new Error('Subscription not found');
    return {
      ...sub,
      userId: 'user-1',
      notificationsEnabled: data.notifications_enabled ?? sub.notificationsEnabled,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };
  }

  const response = await api.patch<Subscription>(`/subscriptions/${subscriptionId}`, data);
  return response.data;
}

/**
 * Unsubscribe from a playlist
 */
export async function deleteSubscription(subscriptionId: string): Promise<void> {
  if (isMock()) {
    return;
  }

  await api.delete(`/subscriptions/${subscriptionId}`);
}

/**
 * List keyword groups linked to a subscription
 */
export async function listLinkedKeywordGroups(subscriptionId: string): Promise<LinkedKeywordGroup[]> {
  if (isMock()) {
    return [
      {
        id: 'link-1',
        keywordGroupId: 'group-1',
        keywordGroupName: 'Home Streets',
        keywordCount: 15,
        createdAt: new Date().toISOString(),
      },
    ];
  }

  const response = await api.get<LinkedKeywordGroup[]>(`/subscriptions/${subscriptionId}/keyword-groups`);
  return response.data;
}

/**
 * Link a keyword group to a subscription
 */
export async function linkKeywordGroup(
  subscriptionId: string,
  data: LinkKeywordGroupRequest
): Promise<LinkedKeywordGroup> {
  if (isMock()) {
    return {
      id: `link-${Date.now()}`,
      keywordGroupId: data.keyword_group_id,
      keywordGroupName: 'Mock Group',
      keywordCount: 0,
      createdAt: new Date().toISOString(),
    };
  }

  const response = await api.post<LinkedKeywordGroup>(
    `/subscriptions/${subscriptionId}/keyword-groups`,
    data
  );
  return response.data;
}

/**
 * Unlink a keyword group from a subscription
 */
export async function unlinkKeywordGroup(subscriptionId: string, groupId: string): Promise<void> {
  if (isMock()) {
    return;
  }

  await api.delete(`/subscriptions/${subscriptionId}/keyword-groups/${groupId}`);
}
